### PR TITLE
Add the ability to block a user after N wrong password attempts

### PR DESCRIPTION
**Pull request by monassfaro**
# What? Why?
Limit the possibility for the user to enter the wrong password to N times.
The user will be blocked after entering wrong passwords N times in a row.

### CODE REVIEW

File: src/main/java/com/icthh/xm/uaa/security/UaaAuthenticationProvider.java [hunk 1]
```diff
@@ -132,6 +134,15 @@ private User getUser(Authentication authentication) {
         return userService.getUser(domainUserDetails.getUserKey());
     }
 
+    private Authentication handleFailedAuthentication(Authentication authentication) {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/security/UaaAuthenticationProvider.java:137]: Pls rename to more describable name

Comment (monassfaro) [reply src/main/java/com/icthh/xm/uaa/security/UaaAuthenticationProvider.java]: Renamed

File: src/main/java/com/icthh/xm/uaa/domain/User.java [hunk 2]
```diff
@@ -216,6 +219,21 @@ public String toString() {
...
+        this.setPasswordAttempts(0);
+    }
+
+    public void updateLastLoginDate() {
+        this.setLastLoginDate(Instant.now());
+    }
+
+    public void incrementPasswordAttempts() {
+        int incrementedPasswordAttempt = this.getPasswordAttempts() + 1;
+
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/domain/User.java:236]: unnecessary enter

File: src/main/java/com/icthh/xm/uaa/service/UserService.java [hunk 3]
```diff
@@ -58,7 +60,6 @@
 @LepService(group = "service.user")
 @Service
 @Transactional
-@AllArgsConstructor
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/service/UserService.java:61]: just change to required argument constructor

File: src/main/java/com/icthh/xm/uaa/service/UserService.java [hunk 4]
```diff
@@ -73,6 +74,31 @@ public class UserService {
     private final UserPermittedRepository userPermittedRepository;
     private final TokenConstraintsService tokenConstraints;
     private final PasswordResetHandlerFactory passwordResetHandlerFactory;
+    private final ApplicationProperties applicationProperties;
+    @Setter(onMethod = @__(@Autowired))
+    private UserService self;
+
+    public UserService(UserRepository userRepository,
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/service/UserService.java:81]: just use required argument constructor annotation

File: src/main/java/com/icthh/xm/uaa/service/UserService.java [hunk 5]
```diff
@@ -650,8 +676,42 @@ public User updateLastLoginDate(User user) {
...
+                    self.disableAccountActivation(user, maxPasswordAttempts);
+                });
+        }
+    }
+
     public void handlePasswordReset(PasswordResetRequest resetRequest) {
         passwordResetHandlerFactory.getPasswordResetHandler(resetRequest.getResetType()).handle(resetRequest);
     }
 
+    @LogicExtensionPoint(value = "DisableAccountActivation")
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/service/UserService.java:711]: this lep should  be invoked only when user.getPasswordAttempts().equals(maxPasswordAttempts) is true

File: src/main/java/com/icthh/xm/uaa/domain/User.java [hunk 6]
```diff
@@ -216,6 +219,21 @@ public String toString() {
...
+    public void resetPasswordAttempts() {
+        this.setPasswordAttempts(0);
+    }
+
+    public void updateLastLoginDate() {
+        this.setLastLoginDate(Instant.now());
+    }
+
+    public void incrementPasswordAttempts() {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/domain/User.java:234]: pls add
```return this```

File: src/main/java/com/icthh/xm/uaa/service/UserService.java (hunk 5 above)
Comment (sergeysenja1992) [reply src/main/java/com/icthh/xm/uaa/service/UserService.java:711]: DisableAccountActivation it's separate lep that should be invoken only when you really disable user, not every failed tries

Comment (sergeysenja1992) [reply src/main/java/com/icthh/xm/uaa/service/UserService.java:711]: and pls rename it to PasswordAttemptsExceeded

File: src/main/java/com/icthh/xm/uaa/domain/User.java (hunk 6 above)
Comment (monassfaro) [reply src/main/java/com/icthh/xm/uaa/domain/User.java]: Added

File: src/main/java/com/icthh/xm/uaa/service/UserService.java (hunk 5 above)
Comment (monassfaro) [reply src/main/java/com/icthh/xm/uaa/service/UserService.java]: Yes, you are absolutely right. Renamed

File: src/main/java/com/icthh/xm/uaa/service/UserService.java (hunk 4 above)
Comment (monassfaro) [reply src/main/java/com/icthh/xm/uaa/service/UserService.java]: Thanks. Fixed

File: src/main/java/com/icthh/xm/uaa/domain/User.java (hunk 2 above)
Comment (monassfaro) [reply src/main/java/com/icthh/xm/uaa/domain/User.java]: Removed redundant line.

File: src/main/java/com/icthh/xm/uaa/domain/User.java [hunk 7]
```diff
@@ -216,6 +219,21 @@ public String toString() {
...
+    public void resetPasswordAttempts() {
+        this.setPasswordAttempts(0);
+    }
+
+    public void updateLastLoginDate() {
+        this.setLastLoginDate(Instant.now());
+    }
+
+    public User incrementPasswordAttempts() {
+        int incrementedPasswordAttempt = this.getPasswordAttempts() + 1;
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/domain/User.java:235]: possible NPE

File: src/main/java/com/icthh/xm/uaa/service/UserService.java [hunk 8]
```diff
@@ -650,8 +656,40 @@ public User updateLastLoginDate(User user) {
...
+        }
+    }
+
     public void handlePasswordReset(PasswordResetRequest resetRequest) {
         passwordResetHandlerFactory.getPasswordResetHandler(resetRequest.getResetType()).handle(resetRequest);
     }
 
+    @LogicExtensionPoint(value = "PasswordAttemptsExceeded")
+    public void passwordAttemptsExceeded(User user, Integer maxPasswordAttempts) {
+        if (user.getPasswordAttempts().equals(maxPasswordAttempts)) {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/service/UserService.java:691]: 1 - move to upside of this method

2 - possible NPE on getPasswordAttempts()

File: src/main/java/com/icthh/xm/uaa/domain/User.java (hunk 7 above)
Comment (monassfaro) [reply src/main/java/com/icthh/xm/uaa/domain/User.java:240]: Handled NPE

**Review by sergeysenja1992 — APPROVED** [review]

**Review by amedvedchuk — APPROVED** [review]

**Pull request merged**